packages/core/admin/admin/src/core/apis/CustomFields.js

```diff
@@ -0,0 +1,32 @@
+import invariant from 'invariant';
+
+class CustomFields {
+  constructor() {
+    this.customFields = {};
+  }
+
+  register(customField) {
+    const { name, pluginId, type, intlLabel, intlDescription, components } = customField;
+
+    // Ensure required attributes are provided
+    invariant(name, 'A name must be provided');
+    invariant(type, 'A type must be provided');
+    invariant(intlLabel, 'An intlLabel must be provided');
+    invariant(intlDescription, 'An intlDescription must be provided');
+    invariant(components, 'A components object must be provided');
+    invariant(components.Input, 'An Input component must be provided');
+
+    // When no plugin is specified, default to the global namespace
+    const namespace = pluginId ? `plugin::${pluginId}.${name}` : `global::global.${name}`;
+
+    // if (this.customFields[namespace] != null) {
+    //   throw Error('already exists');
+    // }
+    // Ensure the namespace is unique
+    invariant(this.customFields[namespace] === undefined, 'A similar custom field already exists');
```

Comment:
I feel like there is probably something I'm missing here. How could this happen if we are already checking for properties on the registered custom field value? Also the message seems a bit weird if the value is undefined.

Reply:
>Also the message seems a bit weird if the value is undefined

I think it's the other way around with invariant: you get the error if the first parameter is falsy, so in this case only if something actually exists. Therefore it throws if a custom field with the same name was already registered

>How could this happen if we are already checking for properties on the registered custom field value?

I'm not sure I understand what you mean here, could you please rephrase?

Reply:
oh gotcha, thanks. I think that clarifies my question. (if a custom field was undefined I think it would throw an error earlier 'cannot read property of undefined' when checking name and type).

But do you need to check the value, aren't you just checking the name is unique on the object. Couldn't you just check the keys? `Object.keys(this.cusomFields).includes(namespace)`

Reply:
Or even `this.customFields.hasOwnProperty(namespace)` then

Reply:
yup! :)